velox/common/caching/SsdCache.h

```diff
@@ -0,0 +1,355 @@
+
+  // Returns [start, size] of contiguous space for storing data of a
+  // number of contiguous 'pins' starting at 'startIndex'.  Returns a
+  // run of 0 bytes if there is no space
+  std::pair<uint64_t, int32_t> getSpace(
+      const std::vector<CachePin>& pins,
+      int32_t begin);
+
+  // Removes all 'entries_' that reference data in regions described by
+  // 'regionIndices'.
+  void clearRegionEntriesLocked(const std::vector<int32_t>& regionIndices);
+
+  // Clears one or more  regions for accommodating new entries. The regions are
+  // added to 'writableRegions_'. Returns true if regions could be cleared.
+  bool evictLocked();
+
+  // Increments event count and periodically decays scores.
+  void newEventLocked();
+
+  // Verifies that 'entry' has the data at 'run'.
+  void verifyWrite(AsyncDataCacheEntry& entry, SsdRun run);
+
+  // Serializes access to all private data members.
+  std::mutex mutex_;
+
+  // The containing SsdCache.
+  SsdCache& cache_;
+
+  // Stripe number within 'cache_'.
+  int32_t ordinal_;
+
+  // Number of kRegionSize regions in the file.
+  int32_t numRegions_{0};
+
+  // True if stopped serving traffic. Happens if no evictions are
+  // possible due to everything being pinned. Clears when pins
+  // decrease and space can be cleared.
+  bool suspended_{false};
+
+  // Maximum size of the backing file in kRegionSize units.
+  const int32_t maxRegions_;
+
+  // Number of used bytes in in each region. A new entry must fit
+  // between the offset and the end of the region.
+  std::vector<uint32_t> regionSize_;
+
+  // Region numbers available for writing new entries.
+  std::vector<int32_t> writableRegions_;
+
+  // Count of bytes read from the corresponding region. Decays with time.
+  std::vector<uint64_t> regionScore_;
+
+  // Pin count for each region.
+  std::vector<int32_t> regionPins_;
+
+  // Map of file number and offset to location in file.
+  folly::F14FastMap<SsdKey, SsdRun> entries_;
+
+  // Count of reads and writes. The scores are decayed every time e count goes
+  // over kDecayInterval or half 'entries_' size, wichever comes first.
+  uint64_t numEvents_{0};
+
+  // Name of backing file.
+  const std::string filename_;
+
+  // File descriptor.
+  int32_t fd_;
+
+  // Size of the backing file in bytes. Must be multiple of kRegionSize.
+  uint64_t fileSize_{0};
+
+  // ReadFile made from 'fd_'.
+  std::unique_ptr<ReadFile> readFile_;
+
+  // Counters.
+  SsdCacheStats stats_;
+};
+
+class SsdCache {
+ public:
+  //  Constructs a cache with backing files at path
+  //  'filePrefix'.<ordinal>. <ordinal> ranges from 0 to 'numShards' -
+  //  1.. '. 'maxBytes' is the total capacity of the cache. This is
+  //  rounded up to the next multiple of kRegionSize * 'numShards'.
+  SsdCache(
+      std::string_view filePrefix,
+      uint64_t maxBytes,
+      int32_t numShards_ = 32);
+
+  // Returns the shard corresponding to 'fileId'.
+  SsdFile& file(uint64_t fileId);
```

Comment:
What is fileId? Do you use shard and file interchangeably? Would it make sense to use only one or the other?